feat(live-status-gw): Add segment duration to `segments` topic

ianshade opened the pull request:
This PR is being opened on behalf of TV 2 Norge.

* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)
A feature: new properties exposed by the Live Status Gateway in the `segments` topic


* **What is the current behavior?** (You can also link to an open issue here)
No durations are available to the subscribers of the `segments` topic.


* **What is the new behavior (if this is a feature change)?**
  New properties are added to `SegmentStatus`:
  - `expectedDurationMs`, calculated as a sum of `expectedDurationWithPreroll` of its parts (when available on at least one part)
  - `budgetDurationMs`, calculated as a sum of `budgetDuration` of its parts (when available on at least one part)

* **Other information**:

**Status**
<!--
Check the checkboxes below as the PR progresses.
The author is encouraged to do a functional test before submitting
-->
- [ ] Code documentation for the relevant parts in the code have been added/updated by the PR author
- [x] The functionality has been tested by the PR author
- [x] Automated tests to cover the new functionality and/or guard against regressions have been added
- [ ] The functionality has been tested by NRK

jstarpl commented:
Hello!

Thank you for contributing to the Sofie Project!

If you haven’t already, please give our [contribution guidelines ](https://nrkno.github.io/sofie-core/docs/for-developers/contribution-guidelines/)a read. Can you write me an e-mail at jan.starzak@nrk.no so that we can schedule a workshop to discuss this change?

jstarpl commented on packages/live-status-gateway/src/topics/helpers/segmentTiming.ts:
```diff
@@ -0,0 +1,50 @@
+			0) + (segmentTiming.budgetDurationMs ?? 0)
+	return {
+		...segmentTiming,
+		expectedEndTime: segmentTiming.budgetDurationMs != null ? expectedBudgetEndTime : expectedEndTime,
```
The meaning of this value can be unclear - see comment on `segments.yaml`

jstarpl commented:
An online workshop to discuss this PR has been scheduled for 2023/11/24 14:00 CET. If you'd like to participate in the workshop, please send an e-mail to jan.starzak@nrk.no so that I can arrange that.

I will publish notes from the workshop in this thread afterwards.

jstarpl commented:
The meeting was held as planned; Mint de Wit and Jan Starzak from the NRK Sofie team and Krzysztof Zegzula representing TV 2 Norge participated.

* The use-case is to build a custom UI to show countdowns to end-of-Part and end-of-Segment, matching what is displayed in the Sofie GUI
  - How should incomplete data be handled?
    - One option is to do the same simulation that Sofie Core GUI does (if a timestamp hasn’t been set yet, use current time).
    - Provide a “special” token (like null) to say that that computation can’t be done yet.
    - We agreed that having a “single” algorithm of fallback for all consumers means consistency at the price of potential greater accuracy, if additional fallback mechanisms are implemented. We agreed that consistency is more valuable.
* Display Duration group handling has a TODO
    - There’s a difference when the 0-`expectedDuration` Part is before the non-zero expectedDuration Part inside
    - Core counts in a cascading fashion, top-to-bottom, while this implementation in Live Status Gateway does a calculation for currentPart only, so it counts all played and unplayed parts the same, no matter their positional relationship to the current Part. This may result in a difference in calculation
* Other code comments from the PR have been discussed and will be addressed
* Timestamp reliability - how should the API note that timestamps are always relative to the Sofie Server Core system clock (assuming synced over NTP)
  - Leaving this as an open question for now, the problem should be noted in the ReadMe for Live Status Gateway (client needs to be in time sync with Live Status Gateway)

ianshade commented on packages/live-status-gateway/src/topics/helpers/partTiming.ts:
```diff
@@ -0,0 +1,35 @@
+	return {
+		startTime: currentPartInstance.timings?.plannedStartedPlayback ?? 0,
+		expectedDurationMs: currentPartInstance.part.expectedDuration ?? 0,
+		expectedEndTime: (currentPartInstance.timings?.plannedStartedPlayback ?? 0) + expectedDuration,
```
This has been addressed in https://github.com/nrkno/sofie-core/pull/1053/commits/eafabf7476bddc5c547bdfbdc4aee47418350b01

ianshade commented on packages/server-core-integration/src/lib/ddpClient.ts:
```diff
@@ -583,6 +583,7 @@ export class DDPClient extends EventEmitter<DDPClientEvents> {
 					delete this.collections[name][id][value]
 				})
 			}
+			this.collections[name][id] = { ...this.collections[name][id] } // allows shallow comparison to detect changed objects
```
I updated this place, and I a similar occurence in the `added` method

ianshade commented on packages/live-status-gateway/api/schemas/activePlaylist.yaml:
```diff
@@ -34,46 +37,86 @@ $defs:
+        description: Unique id of the segment this part belongs to
+        type: string
+      autoNext:
+        description: Should this part progress to the next automatically
```
I updated it to say "If this part will progress to the next automatically" - don't you think that sounds even better?

ianshade commented on packages/live-status-gateway/api/schemas/activePlaylist.yaml:
```diff
@@ -118,3 +161,31 @@ $defs:
+          budgetDurationMs:
+            description: Budget duration of the segment
+            type: number
+          expectedEndTime:
```
`expectedEndTime` is now renamed to `projectedEndTime`

ianshade commented on packages/live-status-gateway/src/topics/helpers/partTiming.ts:
```diff
@@ -0,0 +1,40 @@
+	let expectedDuration = currentPartInstance.part.expectedDuration ?? 0
+
+	if (isMemberOfDisplayDurationGroup && currentPartInstance.part.expectedDuration === 0) {
+		// TODO: support partInstance order edge case
```
Updated to state what is supported.

ianshade commented:
Thank you, Jan and Mint, for your time and valuable insights during the meeting. I believe I have now addressed all comments from the review and brought up during the meeting. I also resolved the merge conflict.
